File: typescript/src/astro/system-name.ts
/**
 * Parsing and formatting of Elite Dangerous procedural system names.
 *
 * A procedural system name has the shape `Region LL-L m[N1-]N2`, e.g.
 * `Synuefe EN-H d11-96` or `Blae Eock KC-C d0`. Its parts are:
 *
 * - **Region** — the sector name (procedural, e.g. `Synuefe`, or hand-authored,
 *   e.g. `Pleiades Sector`).
 * - **L1 L2 L3** — three letters (`EN-H`) giving the boxel's position within the
 *   sector, together with **N1**.
 * - **Mass code** — the size class `a`–`h`.
 * - **N1 / N2** — the boxel's high index and the system's sequence number. When
 *   N1 is zero the game omits it (and its hyphen), so `…d0`, not `…d0-0`.
 *
 * This module only handles the *textual* form. The bit-packed `id64`
 * (system address) lives in `./system-address`.
 *
 * @packageDocumentation
 */

import { canonicalizeSectorName, sectorGridPositionFromName } from './sector-name.js';
import { getHandAuthoredRegionOrigin } from './naming-region-origins.js';
import { assertBoxelCode, packBoxelCode } from './internal/system-name-code.js';

/**
 * The parsed parts of a procedural system name. Letters and mass code are stored
 * as zero-based indices, not characters, so they can feed the address encoder
 * directly.
 */
export interface SystemNameParts {
    /** The region (sector) name, as written — casing is fixed up on canonicalize. */
    readonly regionName: string;
    /** First boxel letter, 0–25 (`A`–`Z`). */
    readonly l1: number;
    /** Second boxel letter, 0–25. */
    readonly l2: number;
    /** Third boxel letter, 0–25. */
    readonly l3: number;
    /** Mass code as a 0–7 size class (`a`–`h`). */
    readonly massCode: number;
    /** High boxel index (the `N1` before the hyphen); 0 when the name omits it. */
    readonly n1: number;
    /** System sequence number (`N2`). */
    readonly n2: number;
}

const CODE_A_UPPER = 'A'.charCodeAt(0);
const CODE_A_LOWER = 'a'.charCodeAt(0);

/**
 * Pack a boxel's `(l1, l2, l3, n1)` letter code into the single base-26 index the
 * address format uses (the **boxel code**).
 *
 * @param l1 - First letter as a zero-based index, `0`–`25` (`A`–`Z`).
 * @param l2 - Second letter as a zero-based index, `0`–`25`.
 * @param l3 - Third letter as a zero-based index, `0`–`25`.
 * @param n1 - The boxel number that follows the letters (`d11-96` → `11`), a
 *   non-negative integer small enough that the packed code stays exact.
 * @returns The packed base-26 boxel code.
 * @throws {RangeError} If any letter is not an integer in 0–25, or `n1` is negative,
 * fractional, or large enough that the packed code would be rounded. Such a field
 * packs as a different boxel than the one asked for rather than failing.
 * @example
 * ```ts
 * import { lettersToBoxelCode } from '@elite-dangerous-almanac/core/astro/system-name';
 *
 * lettersToBoxelCode(4, 13, 7, 11); // the 'EN-H …11' boxel of a d-class sector
 * ```
 */
export function lettersToBoxelCode(l1: number, l2: number, l3: number, n1: number): number {
    return packBoxelCode(l1, l2, l3, n1);
}

/**
 * A boxel's letter code unpacked — the `EN-H …11` half of a procedural name.
 *
 * @remarks
 * The same four fields {@link SystemNameParts} carries, without the region name, mass
 * code and sequence that complete a name.
 *
 * @example
 * ```ts
 * import { boxelCodeToLetters } from '@elite-dangerous-almanac/core/astro/system-name';
 * import type { BoxelLetters } from '@elite-dangerous-almanac/core/astro/system-name';
 *
 * const code: BoxelLetters = boxelCodeToLetters(198_410);
 * // -> { l1: 4, l2: 13, l3: 7, n1: 11 }, the 'EN-H …11' boxel
 * ```
 */
export interface BoxelLetters {
    /** First letter as a zero-based index, `0`–`25`. */
    readonly l1: number;
    /** Second letter as a zero-based index, `0`–`25`. */
    readonly l2: number;
    /** Third letter as a zero-based index, `0`–`25`. */
    readonly l3: number;
    /** The boxel number that follows the letters (`d11-96` → `11`). */
    readonly n1: number;
}

/**
 * Unpack a base-26 boxel code into its `(l1, l2, l3, n1)` letter code — the inverse
 * of {@link lettersToBoxelCode}.
 *
 * @param boxelCode - The packed base-26 boxel code, as `decodeSystemAddress` returns.
 *   Must be a non-negative integer {@link lettersToBoxelCode} could have packed.
 * @returns The three letter indices (`0`–`25` each) and the boxel number `n1`.
 * @throws {RangeError} If `boxelCode` is negative, fractional, or beyond the largest
 * exactly packable code — no such code can be packed, and a negative one would yield
 * letters outside `0`–`25`.
 * @example
 * ```ts
 * import { boxelCodeToLetters, lettersToBoxelCode } from '@elite-dangerous-almanac/core/astro/system-name';
 *
 * boxelCodeToLetters(lettersToBoxelCode(4, 13, 7, 11)); // -> { l1: 4, l2: 13, l3: 7, n1: 11 }
 * ```
 */
export function boxelCodeToLetters(boxelCode: number): BoxelLetters {
    assertBoxelCode(boxelCode);
    return {
        l1: boxelCode % 26,
        l2: Math.trunc(boxelCode / 26) % 26,
        l3: Math.trunc(boxelCode / (26 * 26)) % 26,
        n1: Math.trunc(boxelCode / (26 * 26 * 26)),
    };
}

/**
 * Parse a procedural system name into its parts.
 *
 * The region name may itself contain digits (`Col 285 Sector`), so parsing walks
 * the suffix backwards from the end. Returns `null` for anything that is not a
 * well-formed procedural system name.
 *
 * @remarks
 * Parsing does **not** canonicalize casing: `regionName` is captured verbatim from
 * the input (`parseSystemName('synuefe …').regionName === 'synuefe'`). The letters
 * and mass code become numeric indices, so re-formatting them is always canonical,
 * but the region is not — use {@link canonicalizeSystemName} (or build a
 * {@link ProceduralSystem} via `ProceduralSystem.fromName`) if you need the region re-cased too.
 *
 * @param name - A system name in any casing, e.g. `blae eock kc-c d0`.
 * @returns The parsed parts, or `null` if the name is malformed.
 */
export function parseSystemName(name: string): SystemNameParts | null {
    if (name == null) return null;
    const s = name;
    const lower = s.toLowerCase();
    let i = s.length - 1;

    // The fixed tail " LL-L mN2" is 9 chars, so the whole name needs ≥10 chars
    // (region ≥2), e.g. "Th aa-a a0". Faithful to the EDTS reference bound, which
    // likewise never accepts a 1-char region (none exist in game).
    if (i < 9) return null;
    if (lower[i]! < '0' || lower[i]! > '9') return null;

    while (i > 8 && lower[i]! >= '0' && lower[i]! <= '9') i--;
    const n2 = Number.parseInt(lower.substring(i + 1), 10);
    if (Number.isNaN(n2)) return null;

    let n1 = 0;
    if (lower[i] === '-') {
        i--;
        const vend = i;
        while (i > 8 && lower[i]! >= '0' && lower[i]! <= '9') i--;
        if (i === vend) return null;
        n1 = Number.parseInt(lower.substring(i + 1, vend + 1), 10);
        if (Number.isNaN(n1)) return null;
    }

    if (lower[i]! < 'a' || lower[i]! > 'h') return null; // mass code
    const massCode = lower.charCodeAt(i) - CODE_A_LOWER;
    i--;

    if (lower[i] !== ' ') return null;
    i--;

    if (lower[i]! < 'a' || lower[i]! > 'z') return null; // L3
    const l3 = lower.charCodeAt(i) - CODE_A_LOWER;
    i--;

    if (lower[i] !== '-') return null;
    i--;

    if (lower[i]! < 'a' || lower[i]! > 'z') return null; // L2
    const l2 = lower.charCodeAt(i) - CODE_A_LOWER;
    i--;

    if (lower[i]! < 'a' || lower[i]! > 'z') return null; // L1
    const l1 = lower.charCodeAt(i) - CODE_A_LOWER;
    i--;

    if (lower[i] !== ' ') return null;
    i--;

    return { regionName: s.substring(0, i + 1), l1, l2, l3, massCode, n1, n2 };
}

/**
 * Format parts back into a system name. N1 (and its hyphen) is omitted when zero,
 * matching the in-game rendering.
 *
 * @remarks
 * Letters and mass code are rendered from their numeric indices, so they always
 * come out canonically cased. The `regionName`, however, is emitted **as stored** —
 * `formatSystemName` does not re-case it. Round-tripping a lower-cased name through
 * {@link parseSystemName} → `formatSystemName` therefore keeps the region's original
 * casing (`synuefe EN-H d11-96`); for a fully canonical name use
 * {@link canonicalizeSystemName} or `ProceduralSystem.fromName(...).name`.
 *
 * @param parts - The system-name parts to render.
 * @returns The system name, e.g. `Synuefe EN-H d11-96`.
 */
export function formatSystemName(parts: SystemNameParts): string {
    const l1 = String.fromCharCode(CODE_A_UPPER + parts.l1);
    const l2 = String.fromCharCode(CODE_A_UPPER + parts.l2);
    const l3 = String.fromCharCode(CODE_A_UPPER + parts.l3);
    const mc = String.fromCharCode(CODE_A_LOWER + parts.massCode);
    const n1 = Math.trunc(parts.n1);
    const n2 = Math.trunc(parts.n2);
    const index = n1 !== 0 ? `${n1}-${n2}` : `${n2}`;
    return `${parts.regionName} ${l1}${l2}-${l3} ${mc}${index}`;
}

/**
 * Fix the casing of a system name. Procedural regions are round-tripped through
 * the grid; catalogued hand-authored regions are looked up case-insensitively.
 * Unknown region strings are left untouched so syntactic parsing remains separate
 * from address encodability.
 *
 * @param name - A system name in any casing.
 * @returns The canonically-cased name, or `null` if it is not a system name.
 */
export function canonicalizeSystemName(name: string): string | null {
    const parts = parseSystemName(name);
    if (!parts) return null;
    return formatSystemName({
        ...parts,
        regionName:
            canonicalizeSectorName(parts.regionName) ??
            getHandAuthoredRegionOrigin(parts.regionName)?.name ??
            parts.regionName,
    });
}

/** Options for {@link isProceduralSystemName}. */
export interface IsProceduralSystemNameOptions {
    /**
     * When `true`, also require the region to parse as a procedural sector, so
     * hand-authored region names (e.g. `Pleiades Sector HR-W d1-79`) are rejected.
     * Defaults to `false`.
     */
    readonly strict?: boolean;
}

/**
 * Whether a string is a well-formed procedural system name.
 *
 * @param name - The candidate name.
 * @param options - See {@link IsProceduralSystemNameOptions}.
 * @returns `true` when the name parses as a procedural system name. A hand-named
 * system (`Sol`, `Maia`) is `false`: it is a real system, just not a procedural name.
 * @example
 * ```ts
 * import { isProceduralSystemName } from '@elite-dangerous-almanac/core/astro/system-name';
 *
 * isProceduralSystemName('Blae Eock KC-C d0');                    // -> true
 * isProceduralSystemName('Pleiades Sector HR-W d1-79');           // -> true
 * isProceduralSystemName('Pleiades Sector HR-W d1-79', { strict: true }); // -> false
 * ```
 */
export function isProceduralSystemName(
    name: string,
    options: IsProceduralSystemNameOptions = {},
): boolean {
    const parts = parseSystemName(name?.trim());
    if (!parts) return false;
    return options.strict ? sectorGridPositionFromName(parts.regionName) !== null : true;
}

File: dotnet/src/EliteDangerousAlmanac/Astronomy/SystemName.cs
using System;
using System.Globalization;
using EliteDangerousAlmanac.Astronomy.Internal;

namespace EliteDangerousAlmanac.Astronomy;

/// <summary>The parts one procedural system name is written from.</summary>
/// <param name="RegionName">The region name, as written.</param>
/// <param name="L1">The first boxel letter, zero through 25.</param>
/// <param name="L2">The second boxel letter, zero through 25.</param>
/// <param name="L3">The third boxel letter, zero through 25.</param>
/// <param name="SizeClass">The size class the mass-code letter states, zero through seven.</param>
/// <param name="N1">The boxel index before the hyphen, which is zero where the name omits it.</param>
/// <param name="N2">The system's own number.</param>
/// <remarks>
/// The letters and the mass code are held as counts rather than characters, so they go
/// straight into the address maths.
/// </remarks>
public sealed record SystemNameParts(
    string RegionName,
    int L1,
    int L2,
    int L3,
    int SizeClass,
    int N1,
    int N2);

/// <summary>The four fields one boxel code holds.</summary>
/// <param name="L1">The first letter, zero through 25.</param>
/// <param name="L2">The second letter, zero through 25.</param>
/// <param name="L3">The third letter, zero through 25.</param>
/// <param name="N1">The number after the letters.</param>
public sealed record BoxelLetters(int L1, int L2, int L3, int N1);

/// <summary>The written form of a procedural system name.</summary>
/// <remarks>
/// <para>
/// A procedural system name reads as <c>Region LL-L m[N1-]N2</c>, as in
/// <c>Synuefe EN-H d11-96</c> or <c>Blae Eock KC-C d0</c>. The region is the sector name,
/// procedural or hand-written. The three letters and the first index state the boxel
/// inside that region, the mass code states how large the boxel is, and the last index
/// numbers the system inside it. The game leaves the first index and its hyphen out where
/// it is zero.
/// </para>
/// <para>
/// This class reads and writes the text alone. <see cref="SystemAddress"/> holds the
/// packed form the game identifies a system by.
/// </para>
/// </remarks>
public static class SystemName
{
    /// <summary>The shortest name the grammar accepts, as in <c>Th aa-a a0</c>.</summary>
    /// <remarks>
    /// The fixed ending takes nine characters, and the shortest region the game writes
    /// takes two. The EDTS reference draws the bound in the same place.
    /// </remarks>
    private const int ShortestName = 10;

    /// <summary>Packs a boxel's letters and index into one base-26 code.</summary>
    /// <param name="l1">The first letter, zero through 25.</param>
    /// <param name="l2">The second letter, zero through 25.</param>
    /// <param name="l3">The third letter, zero through 25.</param>
    /// <param name="n1">The number after the letters.</param>
    /// <returns>The packed code.</returns>
    /// <exception cref="ArgumentOutOfRangeException">
    /// A letter is outside zero through 25, or the number is too large to pack exactly.
    /// </exception>
    public static int ToBoxelCode(int l1, int l2, int l3, int n1) =>
        BoxelCodes.Pack(l1, l2, l3, n1);

    /// <summary>Takes the letters and the index back out of one boxel code.</summary>
    /// <param name="boxelCode">The packed code, as the address decoder answers.</param>
    /// <returns>The three letters and the number after them.</returns>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The code is not one <see cref="ToBoxelCode"/> answers.
    /// </exception>
    public static BoxelLetters ToBoxelLetters(int boxelCode)
    {
        BoxelCodes.Require(boxelCode, nameof(boxelCode));
        return new BoxelLetters(
            boxelCode % 26,
            boxelCode / 26 % 26,
            boxelCode / (26 * 26) % 26,
            boxelCode / (26 * 26 * 26));
    }

    /// <summary>Reads a procedural system name into its parts.</summary>
    /// <param name="name">The name in any casing, with any surrounding whitespace.</param>
    /// <returns>
    /// The parts, or <see langword="null"/> where the name is not a procedural system name.
    /// An absent name is a miss, so a caller hands over a field an import may not carry.
    /// </returns>
    /// <remarks>
    /// A region name carries digits of its own, as <c>Col 285 Sector</c> does, so the
    /// reading walks the ending backwards. The letters and the mass code come out as
    /// counts, and therefore write back in the game's own casing. The region comes out as
    /// written; <see cref="Canonicalize"/> re-cases that too.
    /// </remarks>
    public static SystemNameParts? Parse(string? name)
    {
        if (name is null) return null;
        string trimmed = name.Trim();
        string lower = trimmed.ToLowerInvariant();
        int at = trimmed.Length - 1;

        if (trimmed.Length < ShortestName) return null;
        if (!IsDigit(lower[at])) return null;

        while (at > 8 && IsDigit(lower[at])) at--;
        if (!TryReadNumber(lower, at + 1, trimmed.Length - 1, out int n2)) return null;

        int n1 = 0;
        if (lower[at] == '-')
        {
            at--;
            int end = at;
            while (at > 8 && IsDigit(lower[at])) at--;
            if (at == end) return null;
            if (!TryReadNumber(lower, at + 1, end, out n1)) return null;
        }

        if (lower[at] < 'a' || lower[at] > 'h') return null;
        int sizeClass = lower[at] - 'a';
        at--;

        if (lower[at] != ' ') return null;
        at--;

        if (!IsLetter(lower[at])) return null;
        int l3 = lower[at] - 'a';
        at--;

        if (lower[at] != '-') return null;
        at--;

        if (!IsLetter(lower[at])) return null;
        int l2 = lower[at] - 'a';
        at--;

        if (!IsLetter(lower[at])) return null;
        int l1 = lower[at] - 'a';
        at--;

        if (lower[at] != ' ') return null;
        at--;

        return new SystemNameParts(trimmed.Substring(0, at + 1), l1, l2, l3, sizeClass, n1, n2);
    }

    /// <summary>Writes parts back out as a system name.</summary>
    /// <param name="parts">The parts to write.</param>
    /// <returns>The name, such as <c>Synuefe EN-H d11-96</c>.</returns>
    /// <exception cref="ArgumentNullException">The parts are absent.</exception>
    /// <exception cref="ArgumentOutOfRangeException">
    /// A letter is outside zero through 25, or the size class is outside zero through seven.
    /// </exception>
    /// <remarks>
    /// The first index and its hyphen are left out where the index is zero, which is what
    /// the game does. The region is written as it is held, so a name read in lower case
    /// keeps its own region casing.
    /// </remarks>
    public static string Format(SystemNameParts parts)
    {
        if (parts is null) throw new ArgumentNullException(nameof(parts));

        char l1 = Letter(parts.L1, nameof(parts));
        char l2 = Letter(parts.L2, nameof(parts));
        char l3 = Letter(parts.L3, nameof(parts));
        char massCode = MassCode.FromSizeClass(parts.SizeClass);
        string index = parts.N1 != 0
            ? string.Format(CultureInfo.InvariantCulture, "{0}-{1}", parts.N1, parts.N2)
            : parts.N2.ToString(CultureInfo.InvariantCulture);

        return string.Format(
            CultureInfo.InvariantCulture,
            "{0} {1}{2}-{3} {4}{5}",
            parts.RegionName,
            l1,
            l2,
            l3,
            massCode,
            index);
    }

    /// <summary>Writes a system name in the game's own casing.</summary>
    /// <param name="name">The name in any casing, with any surrounding whitespace.</param>
    /// <returns>
    /// The name as the game writes it, or <see langword="null"/> where it is not a system
    /// name.
    /// </returns>
    /// <remarks>
    /// A procedural region is re-cased by reading it through the grid, and a region the
    /// game names by hand by looking it up. A region neither answers is left as written, so
    /// reading a name stays apart from addressing one.
    /// </remarks>
    public static string? Canonicalize(string? name)
    {
        SystemNameParts? parts = Parse(name);
        if (parts is null) return null;

        string region = SectorName.Canonicalize(parts.RegionName)
            ?? NamingRegionOrigins.FindHandAuthored(parts.RegionName)?.Name
            ?? parts.RegionName;
        return Format(parts with { RegionName = region });
    }

    /// <summary>Reads whether a name is a well-formed procedural system name.</summary>
    /// <param name="name">The name in any casing, with any surrounding whitespace.</param>
    /// <param name="requireProceduralRegion">
    /// Whether the region must itself be a procedural sector. Where it is
    /// <see langword="true"/>, a system inside a region the game names by hand, such as
    /// <c>Pleiades Sector HR-W d1-79</c>, is not one.
    /// </param>
    /// <returns>
    /// <see langword="true"/> where the name reads as a procedural system name. A system the
    /// game names outright, such as <c>Sol</c>, is not one. An absent name is not one either.
    /// </returns>
    public static bool IsProcedural(string? name, bool requireProceduralRegion = false)
    {
        SystemNameParts? parts = Parse(name);
        if (parts is null) return false;
        return !requireProceduralRegion || SectorName.ToGridPosition(parts.RegionName) is not null;
    }

    /// <summary>Reads whether one character is a digit.</summary>
    private static bool IsDigit(char value) => value >= '0' && value <= '9';

    /// <summary>Reads whether one character is a lower-case letter.</summary>
    private static bool IsLetter(char value) => value >= 'a' && value <= 'z';

    /// <summary>Reads the number one run of digits states.</summary>
    private static bool TryReadNumber(string value, int start, int end, out int number)
    {
        number = 0;
        if (end < start) return false;
        return int.TryParse(
            value.AsSpan(start, end - start + 1),
            NumberStyles.None,
            CultureInfo.InvariantCulture,
            out number);
    }

    /// <summary>Writes one letter count as its letter.</summary>
    private static char Letter(int index, string parameter)
    {
        if (index < 0 || index > 25)
        {
            throw new ArgumentOutOfRangeException(
                parameter, index, "A boxel letter is an integer from 0 through 25.");
        }

        return (char)('A' + index);
    }
}
